src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/AbstractBuildEventValueProvider.AbstractBuildEventHelper.cs

```diff
@@ -0,0 +1,189 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Build.Construction;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Properties.InterceptedProjectProperties
+{
+    internal abstract partial class AbstractBuildEventValueProvider
+    {
+        public abstract class AbstractBuildEventHelper
+        {
+            private const string _execTask = "Exec";
+            private const string _command = "Command";
+
+            protected AbstractBuildEventHelper(string buildEvent,
+                   string targetName,
+                   Func<ProjectTargetElement, string> getTarget,
+                   Action<ProjectTargetElement> setTargetDependencies)
+            {
+                BuildEvent = buildEvent;
+                TargetName = targetName;
+                GetTarget = getTarget;
+                SetTargetDependencies = setTargetDependencies;
+            }
+
+            private Func<ProjectTargetElement, string> GetTarget { get; }
+            private Action<ProjectTargetElement> SetTargetDependencies { get; }
+            private string BuildEvent { get; }
+            private string TargetName { get; }
+
+            public string GetProperty(ProjectRootElement projectXml)
+            {
+                // Check if project file already has props in place for this.
+                var result = TryGetFromProperty(projectXml);
```

Comment:
I don't like the pattern of a method returning null to indicate failure.  `null` is a valid thing to return from `GetProperty` itself.  Someone could mistakenly assume that they should return null if one of the methods they call gives them null, instead of continuing.  Roslyn uses [Optional](http://source.roslyn.io/#Microsoft.CodeAnalysis/Optional.cs,86bbd87a0ffc79bf,references) to indicate whether a returned value is valid, but this pattern was established before tuples.